### DISCUSSION TITLE
Until the new system is out can someone help me add brands taxonomy terms to new product in REST?

### DISCUSSION THREAD

**Discussion by rudtek** [woocommerce/woocommerce#52469] [Brands]
I created my own custom brand (product_brands) as a taxonomy and added it to the REST API using this code:

```
add_filter( 'woocommerce_rest_prepare_product_object', 'ul_rest_api_prepare_brands_tax', 10, 3 );
function ul_rest_api_prepare_brands_tax( $response, $object, $request ) {
    $product_id = $object->get_id();
    if ( empty( $response->data[ 'product_brands' ] ) ) {
        $terms = [];
        foreach ( wp_get_post_terms( $product_id, 'product_brands' ) as $term ) {
            $terms[] = [
                'id'   => $term->term_id,
                'name' => $term->name,
                'slug' => $term->slug,
            ];
        }
        $response->data[ 'product_brands' ] = $terms;
    }
    return $response;
}
```
I can see brands attached when I use a GET for products, and with only this code I could not PUT OR POST a product with this brand.  

So I added this code:

```
function ul_prepare_insert_product( $product, $request ) {
    if ( isset( $request['product_brands'] ) ) {
        wp_set_post_terms($request['id'], $request['product_brands'], 'product_brands', false);
    }
    return $product;
}
add_filter( 'woocommerce_rest_pre_insert_product_object', 'ul_prepare_insert_product', 10, 2 );
```

This works if I PUT an array to the body of an existing product endpoint to allow me to add brand terms.  My array is just `product_brands[61]`  (which is the ID of one of my existing brand terms.) I found that I can use an ID if I use `wp_set_post_terms` (not `wp_set_object_terms`).  `set_object_terms` will only allow me to use the brand slug.

However, neither option works if I try to POST a NEW product with an existing brand term.  Nothing gets set into the brand for the product. 

I've looked high and low for documentation on this but came up short.  Here's some further notes/assumptions and questions.

Some sites said to use `woocommerce_rest_pre_insert_product`  and some said `woocommerce_rest_pre_insert_product_object`.  My understanding is that using `object` is the new way to do it.

I didn't get some of the parameters either. Sometimes it was `$object` and sometimes it was `$product`. Either case seems to work but I think it may have to do with whether I used the filter `woocommerce_rest_pre_insert_product_object` or `woocommerce_rest_pre_insert_post_object`. I'm not sure.

**Comment by danielortega-dev** [comment]
Thanks for your post.
So, I will take look at the code within 10 mins

**Reply by rudtek** [reply]
thank you. I saw this post too, which may or may not help.
[Stackoverflow custom taxonomy](https://stackoverflow.com/questions/76744011/cannot-assign-custom-taxonomy-to-woocommerce-product-via-rest-api?rq=2)

**Reply by danielortega-dev** [reply]
`woocommerce_rest_pre_insert_product_object` Filter: This filter is correct for both `POST` and `PUT` requests when dealing with product objects.

```
add_filter('woocommerce_rest_pre_insert_product_object', 'ul_prepare_insert_product', 10, 2);
function ul_prepare_insert_product($product, $request) {
    if (isset($request['product_brands'])) {
        wp_set_object_terms($product->get_id(), $request['product_brands'], 'product_brands');
    }
    return $product;
}
```

please use this code

**Comment by rudtek** [comment]
Thanks for helping me on this.  That's the same code that I had but using `wp_set_object` instead of `wp_set_post` This code does work when I use a PUT and the body has the array `product_brands[61]`  it does not work for POST though at all.  my results show an empty array for `product_brands`.

my array that I'm sending when I PUT is:
```
{
	"name": "Beaut2",
	"type": "simple",
	"regular_price": "21.99",
	"description": "Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Vestibulum tortor quam, feugiat vitae, ultricies eget, tempor sit amet, ante. Donec eu libero sit amet quam egestas semper. Aenean ultricies mi vitae est. Mauris placerat eleifend leo.",
	"short_description": "Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas.",
	"categories": [
	{
		"id": 9
	},
	{
		"id": 14
	}],
	"product_brands": [
		62
    ]
}
```

**Reply by danielortega-dev** [reply]
Can we chat on Slack?
